Context position features

vgkz opened the pull request:
Adds 2 scripts to get positional features from betalab and context sequences from corpus.

vgkz committed:
add script to create positional features

vgkz committed:
add script to create context sequences

vgkz committed:
change default max_length to 120

vgkz committed:
update get_positional_features() to handle page offsets

vgkz commented:
Currently get_positional_features.py generates missing data if a page does not exist on betalab. The code matches text blocks on betalab to text sequences in the xml files so if a text sequence does not correspond to a text block on betalab there will also be missing data.

ninpnin commented on src/get_context_sequences.py:
```diff
@@ -0,0 +1,215 @@
+def main(args):
+    
+    # chose type of context sequence to output
+    if args.context_type == 'left_context':
```
Replace with this https://stackoverflow.com/a/15836714/12764816

ninpnin commented on src/get_context_sequences.py:
```diff
@@ -0,0 +1,215 @@
+        raise Exception(f'{args.context_type} is an invalid context type. Valid inputs are: [left_context, full_context]')
+    
+    protocols = sorted(list(protocol_iterators(args.records_folder, start=args.start, end=args.end)))
+    curr_year = protocols[0].split('\\')[-2]
```
use pyriksdagen.utils infer_metadata

ninpnin commented on src/get_context_sequences.py:
```diff
@@ -0,0 +1,215 @@
+        next_year = protocol.split('\\')[-2]
+        if curr_year != next_year:
+            output_df = pd.DataFrame.from_dict(context_sequence_dict)
+            save_file = curr_year + '_' + args.context_type + '.csv'
```
We want to avoid saving to a CSV altogether, but that's a worry for another day.

ninpnin commented on src/get_context_sequences.py:
```diff
@@ -0,0 +1,215 @@
+import pandas as pd
+
+####    utility functions   ####
+def cleanup_segments(text_seq):
```
Using .split() without an argument should split on all whitespace, no need to do the translate part.

ninpnin commented on src/get_context_sequences.py:
```diff
@@ -0,0 +1,215 @@
+    text_seq = text_seq.translate({ord('\n'): ' '})
+    text_seq = text_seq.split(' ')
+    text_seq_list = [s for s in text_seq if s != '']
+    text_seq_string = ' '.join(text_seq_list).strip()
```
strip is unnecessary if the value comes directly from a join

ninpnin commented on src/get_context_sequences.py:
```diff
@@ -0,0 +1,215 @@
+    return text_seq_string
+
+####    functions to create context sequences from multiple strings ####
+def concat_sequences_left(previous_sequence, current_sequence, target_length = 120):
```
Doing this is equivalent to doing concat_sequences_full, right? (maybe with a longer max length). 

We should try and remove one of these functions, and then apply an operation to skip the context after. This removes surface area and makes it easier to maintain and debug.

However, when you make this change, run the original code and check that you get _exactly_ the same result after doing it. Otherwise our training data no longer matches the data we apply this on.

ninpnin commented on src/get_context_sequences.py:
```diff
@@ -0,0 +1,215 @@
+        prev_last_sentence_as_list = prev_last_sentence_as_list[-max_previous_length:]
+        prev_last_sentence = ' '.join(prev_last_sentence_as_list)
+    # use new line (/n) as token to signify where current sequence begings
+    return prev_last_sentence + ' /n ' + current_sequence + ' /n ' + next_first_sentence
```
"\n" is the standard symbol. However, you have a special token for this, right? Then it doesn't really matter and we keep it this way.

ninpnin replied:
I don't think it's necessary right now. There are no other "/"'s in the data anyway, so it does not make much of a difference.

But we can try it later.

ninpnin commented on src/get_context_sequences.py:
```diff
@@ -0,0 +1,215 @@
+                   'context_sequence' : context_sequence_list}
+    return output_dict
+
+def get_context_sequence_full(protocol, max_length = 120):
```
These two functions are also pretty much the same.

ninpnin commented on src/get_context_sequences.py:
```diff
@@ -0,0 +1,215 @@
+    for tag, elem in elem_iter(root):
+        if tag == 'note':
+            elem_sequence = elem.text
+            elem_idx = elem.attrib['{http://www.w3.org/XML/1998/namespace}id']
```
pyriksdagen.utils.XML_NS

ninpnin commented on src/get_context_sequences.py:
```diff
@@ -0,0 +1,215 @@
+    parser = etree.XMLParser(remove_blank_text=True)
+    root = etree.parse(protocol, parser).getroot()
+    
+    prev_idx = False
```
Misleading variable name (?)

ninpnin commented on src/get_positional_features.py:
```diff
@@ -0,0 +1,499 @@
+    n_colons_in_curr_sequence = 0
+    
+    # loop through elements on page
+    for elem in xml_tree.iter():
```
See documentation https://envinorma.github.io/alto/alto.html

vgkz committed:
change argparse --context_type to use choices argument rather than throwing an error if wrong context type is given

ninpnin commented on src/get_positional_features.py:
```diff
@@ -0,0 +1,499 @@
+    else:
+        return 'unicameral'
+    
+def second_chamber_bin(chamber):
```
Replace with

```py
>>> [1 if chamber == "ek" else 0  for chamber in ["ek", "ak", "fk"]]
[1, 0, 0]
```

ninpnin commented on src/get_positional_features.py:
```diff
@@ -0,0 +1,499 @@
+                break
+    return pkg, pkg_id
+
+def page_as_string(page_number):
```
https://stackoverflow.com/a/24386708/12764816

ninpnin commented on src/get_positional_features.py:
```diff
@@ -0,0 +1,499 @@
+        feature_dict['page_offset'].extend(protocol_feature_dict['page_offset'])
+    
+    # store features in dataframe and save to disk
+    save_file = curr_year + '_position_features.csv'
```
If we do want to save the thing as a CSV; you should definitely have it in the args. Otherwise we will never know where the file is.

vgkz committed:
switch to using infer_metadata() from pyriksdagen.utils to get protocol year

vgkz committed:
update cleanup_segments() to remove unecessary code
combine context_sequence functions for left context and full context into one function
add pyriksdagen.utils.XML_NS
specify that relative_page_number() outputs floats
change method to compute even page in even_or_odd()
rename misleading var name prev_idx to prev_elem_is_text_seq

vgkz committed:
switch to getting pkg_name from protocol preface
simplify function to get pagenumber in string format

vgkz committed:
add save_folder arg to argparse

vgkz committed:
functions moved to pyriksdagen.utils, change main() script to use imported function from pyriksdagen.utils

vgkz committed:
remove functions to get chamber and replace with infer_metadata and list comprehension

vgkz committed:
change to calling infer_metadata once instead of twice

vgkz committed:
switch to using alto module to parse alto files

vgkz committed:
add script to download alto_files from kblab and save locally

vgkz committed:
update get_positional_features.py to use local alto folder to get alto files
change get_page_pos() to produce less missing data

vgkz committed:
fix: change to correct year when creating new directory

vgkz committed:
remove kblab archive and move args.alto_folder to function variable

vgkz committed:
get data from kblab if no alto-folder is provided

BobBorges commented:
https://github.com/swerik-project/riksdagen-records/issues/104
